operator/internal/handlers/lokistack_create_or_update.go

```diff
@@ -0,0 +1,189 @@
+package handlers
+
+import (
+	"context"
+	"fmt"
+	"os"
+
+	"github.com/ViaQ/loki-operator/internal/manifests/openshift"
+
+	"github.com/ViaQ/logerr/kverrors"
+	"github.com/ViaQ/logerr/log"
+	lokiv1beta1 "github.com/ViaQ/loki-operator/api/v1beta1"
+	"github.com/ViaQ/loki-operator/internal/external/k8s"
+	"github.com/ViaQ/loki-operator/internal/handlers/internal/gateway"
+	"github.com/ViaQ/loki-operator/internal/handlers/internal/secrets"
+	"github.com/ViaQ/loki-operator/internal/manifests"
+	"github.com/ViaQ/loki-operator/internal/metrics"
+	"github.com/ViaQ/loki-operator/internal/status"
+
+	corev1 "k8s.io/api/core/v1"
+	rbacv1 "k8s.io/api/rbac/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/runtime"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// CreateOrUpdateLokiStack handles LokiStack create and update events.
+func CreateOrUpdateLokiStack(ctx context.Context, req ctrl.Request, k k8s.Client, s *runtime.Scheme, flags manifests.FeatureFlags) error {
+	ll := log.WithValues("lokistack", req.NamespacedName, "event", "createOrUpdate")
+
+	var stack lokiv1beta1.LokiStack
+	if err := k.Get(ctx, req.NamespacedName, &stack); err != nil {
+		if apierrors.IsNotFound(err) {
+			// maybe the user deleted it before we could react? Either way this isn't an issue
+			ll.Error(err, "could not find the requested loki stack", "name", req.NamespacedName)
+			return nil
+		}
+		return kverrors.Wrap(err, "failed to lookup lokistack", "name", req.NamespacedName)
+	}
+
+	img := os.Getenv(manifests.EnvRelatedImageLoki)
+	if img == "" {
+		img = manifests.DefaultContainerImage
+	}
+
+	gwImg := os.Getenv(manifests.EnvRelatedImageGateway)
+	if gwImg == "" {
+		gwImg = manifests.DefaultLokiStackGatewayImage
+	}
+
+	var s3secret corev1.Secret
+	key := client.ObjectKey{Name: stack.Spec.Storage.Secret.Name, Namespace: stack.Namespace}
```

Comment:
Do we support only s3 ?

Reply:
For now yes, we have only S3 support. But there is a pending issue to expose more than S3: https://github.com/ViaQ/loki-operator/issues/103